docs/specs/alcproxies.md

```diff
@@ -0,0 +1,251 @@
+Let's say you want to call method `M()` on your `TargetObject`, which returns an object of type `A`, a non-primitive. When it returns the object from the target ALC (where `A` was created) back to the user ALC (where we made the original call for `M()`), we've created an extra reference between the ALCs that we need to deal with. There are a few ways we could try and fix this issue:
+* Don't support non-primitive objects being passed to and from the ALCs
+This isn't recommended since it limits use of the target ALC by quite a lot, but for a quick and dirty version, it could work.
+* For each non-primitive object, generate another proxy that passes back to the user ALC instead of the reference to the normal object. When we unload, we can then remove this proxy as well, keeping the isolation barrier intact.
+This is very similar to how AppDomains handle MarshallByRef objects, and it keeps the functionality the same, but also adds a lot more overhead from creating the proxies during runtime for every non-primitive object.
+* Keep everything call-by-value, by making sure any `ServerObject` returns to the user ALC is the metadata of any object, that gets reconstructed when returning to `ClientObject`
+This option is nice since it works for both inter-process and in-process communication. However, having every object only able to return by value means that we'll be limited to making calls changing objects to a single ALC, and the multiple copies of an object may increase space and cause performance issues.
+
+For my planned implementation, I'll be trying to implement call-by-value, since it keeps the design consistent for all types (primitive and non-primitive) and for in-proc vs cross-proc. However, due to the design for extensibility, it's possible that anyone who wants to change the API can implement any of these options in their own way, by changing the interactions of `ClientObject` and `ServerObject`. If someone wants to limit interactions between plugins and the main context to primitives only, they're welcome to do so with this design.
+
+#### Generic Types
+For generics, the API should probably be able to allow users to pass in a list of types "in order" that are being used for the particular instance of an object they are creating. 
+
+`DispatchProxy` should be OK to use generic types, but it may be an issue funneling these types through whatever server structure is created. 
+
+For constraints on certain generics, the compiler should pick up any issues with constraints from the given interface, but may not from constraints for the TargetObject, and will throw an error.
+
+#### Performance
+No matter the performance of this API, by design it won't have the broader performance issues that AppDomains had in .Net Framework, since they work very similarly to normal classes. However, there may be some overhead and performance issues on any objects being contacted by the proxies, which should be investigated if they come up.
```

Comment:
I think it's actually quite the opposite. However we implement this we're unlikely to get to the perf of remoting on .NET Framework. The main problem will be the usage of `Method.Invoke` for everything - this is known to be pretty slow. There might be ways to optimize this (by generating code at runtime), the question is by how much. In general my expectation would be that this is going to be slower... unless we make significant investment into it in some way...

Reply:
@vitek-karas, if you use [MethodInfo.CreateDelegate](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.methodinfo.createdelegate?view=netcore-2.2) and cache the delegate to call, you don't pay the cost of Method.Invoke. A lot of the cost with Method.Invoke is checking the parameters match the method signature. CreateDelegate checks the prototype of the delegate matches once creates a strongly typed delegate to call which skips the type checking as the .NET type system can now enforce it.

Reply:
Thank @mconnew, that's a great tip... 
I still suspect that the way .NET Framework did remoting was at least on par with the create delegate approach if not better, but obviously we should measure that to be sure.